core/src/main/scala/com/databricks/labs/remorph/preprocessor/README.md

```diff
@@ -0,0 +1,205 @@
+will merely pass through what it sees. However, there may well be other functionality for the preprocessor
+to provide in the future such as perhaps parameter tracking/processing, or perhaps daisy-chaining of
+pre-processors that perform specific tasks.
+
+Hence the preprocessor will be a simple text processor that will:
+
+ - find and replace all `{{ }}`, `{# #}` and `{% %}` blocks with a placeholder
+ - track the unique placeholders and their original text
+ - pass through the placeholders to the target SQL lexer
+ - modify the grammars such that the placeholders are accepted at strategic points in the parse
+ - attempt to convert any SQL within expression macros (complicated by the fact that the Jinga 
+   syntax contains things that look like function calls - we are likely to need a Jinga parser
+ - attempt to convert the SQL within the text where the macros are replaced with placeholders.
+ - the code generator will just pass the placeholders back to the template processor and 
+   the template processor will replace the placeholders with the translated templates.
+
+## Workflow
+```mermaid
+---
+title: Preprocessor Workflow
+config:
+  direction: RL
+  theme: dark
+---
+flowchart TD
+    textsource[Raw Template Source Code] --> preprocessor-phase-I
+    subgraph preprocessor-phase-I[Prepare Input]
+       template-locating[Template locating] --> antlr-char-stream[Text Streaming]
+       antlr-char-stream --> preprocessor-lexer[Tokenizer]
+       preprocessor-lexer --> capture-templates[Capture Templates]
+       capture-templates --> translate[Translate dialect]
+       translate --> template-store[Template Store]
+       preprocessor-lexer --> text-with-placeholders[Text with Placeholders]
+       text-with-placeholders --> dialect-transpiler[Translate Dialect]
+    end
+   template-store --> preprocessor-phase-II
+   dialect-transpiler --> preprocessor-phase-II
+    
+    subgraph preprocessor-phase-II[Translate Templates]
+       template-replacer[Replace Template Placeholders] --> code-gen[Assemble output]
+    end
+    code-gen --> output[Translated Template]
+```
+
+## DBT Projects
+
+As well as the template processing, we will need to bring in the DBT configuration and associated fluff that
+goes with a complete DBT project layout. This includes processing the .yml configuration files and using
+anything from the configuration that is required to generate the Databricks DBT project. This will be a separate
+utility which will coordinate with the preprocessor to ensure that the correct configuration is used and that
+the translated templates are stored in the appropriate location for the DBT project.
+
+A simplified sequence diagram of the process is shown below:
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant DBTProcessor
+    participant ConfigLoader
+    participant Preprocessor
+    participant Template Directory
+    participant OutputWriter
```

Comment:
make references to the existing components we have.

I also think we should create a copy of a dbt project instead of doing in-place translation... though, it might be a configurable option.